solidity/ecdsa/contracts/WalletRegistry.sol

```diff
@@ -0,0 +1,284 @@
+
+        dkg.lockState();
+        dkg.start(
+            uint256(keccak256(abi.encodePacked(relayEntry, block.number)))
+        );
+    }
+
+    /// @notice Submits result of DKG protocol. It is on-chain part of phase 14 of
+    ///         the protocol. The DKG result consists of result submitting member
+    ///         index, calculated group public key, bytes array of misbehaved
+    ///         members, concatenation of signatures from group members,
+    ///         indices of members corresponding to each signature and
+    ///         the list of group members.
+    ///         When the result is verified successfully it gets registered and
+    ///         waits for an approval. A result can be challenged to verify the
+    ///         members list corresponds to the expected set of members determined
+    ///         by the sortition pool.
+    ///         A candidate wallet is registered based on the submitted DKG result
+    ///         details.
+    /// @dev The message to be signed by each member is keccak256 hash of the
+    ///      calculated group public key, misbehaved members indices and DKG
+    ///      start block. The calculated hash should be prefixed with prefixed with
+    ///      `\x19Ethereum signed message:\n` before signing, so the message to
+    ///      sign is:
+    ///      `\x19Ethereum signed message:\n${keccak256(groupPubKey,misbehavedIndices,startBlock)}`
+    /// @param dkgResult DKG result.
+    function submitDkgResult(DKG.Result calldata dkgResult) external {
+        dkg.submitResult(dkgResult);
+    }
+
+    /// @notice Notifies about DKG timeout. Pays the sortition pool unlocking
+    ///         reward to the notifier.
+    function notifyDkgTimeout() external {
+        dkg.notifyTimeout();
+
+        // TODO: Implement transferDkgRewards
+        // transferDkgRewards(msg.sender, sortitionPoolUnlockingReward);
+
+        dkg.complete();
+    }
+
+    /// @notice Approves DKG result. Can be called when the challenge period for
+    ///         the submitted result is finished. Considers the submitted result
+    ///         as valid, pays reward to the approver, bans misbehaved group
+    ///         members from the sortition pool rewards, and completes the group
+    ///         creation by activating the candidate group. For the first
+    ///         `resultSubmissionEligibilityDelay` blocks after the end of the
+    ///         challenge period can be called only by the DKG result submitter.
+    ///         After that time, can be called by anyone.
+    /// @dev It transfers Wallet's ownership to the Wallet Factory owner.
+    /// @param dkgResult Result to approve. Must match the submitted result
+    ///        stored during `submitDkgResult`.
+    function approveDkgResult(DKG.Result calldata dkgResult) external {
+        uint32[] memory misbehavedMembers = dkg.approveResult(dkgResult);
+
+        uint256 walletID = wallets.addWallet(
```

Comment:
What will stop from adding a new wallet in case dkg result was challenged and proved to be invalid? In the prev. approach a wallet was deleted during the challenge step, so there was nothing to add. In this approach I believe we need to have a flag set in `challengeDkgResult` which indicates that dkg was invalid, and another one in `submitDkgResult` to reset that flag. Or some other approach to prevent from adding a new wallet in case of an invalid dkg result.

Reply:
Yes, but the way `challengeDkgResult` worked is that it removed a group/wallet when it was proven to be invalid. In this case, we add a wallet to the mapping at the last step, after the `challegeDkgResult`, which means there's nothing for `challegeDkgResult` to remove. Say a challenger proved the dkg was invalid, but I don't see anything stopping another caller to call `approveDkgResult` with the invalid dkg result. 

Another approach might be reading `dkg.submittedResultHash` which is set during submission and is reset during `challegeDkgResult` when it's proven to be invalid. So maybe adding something like this would do the trick:
```
function approveDkgResult(DKG.Result calldata dkgResult) external {
  require(dkg.submittedResultHash() != '', "DKG result must be submitted and valid");
  (...)
```

Reply:
Currently, in `challengeDkgResult` we call `dkg.challengeResult` which cleans up references to a submitted DKG result and resets the state of DKG to be awaiting a DKG result submission. It's not possible to approve a DKG result without submitting a new result. `approveDkgResult` function calls `dkg.approveResult` which will revert.

Reply:
You're right, this should stop approving a result, after the `dkg.challengeResult` cleans the references.

```
require(
  keccak256(abi.encode(result)) == self.submittedResultHash,
  "Result under approval is different than the submitted one"
);
```